packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -1,3 +1,5 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
```

Comment:
@Brechtpd have you referenced to any third party theory or code for these gadgets in this file? It may be helpful to know which ones are from other people's ideas and which ones are your original.

Reply:
Lots of the low level ones are from the link I posted a couple of days ago: https://www.youtube.com/watch?v=Uug5p05_wqs (with pdf: https://github.com/daira/r1cs/blob/master/zkproofs.pdf).

I think most of the gadgets are written with the help of the basic gadgets that are described above, so no direct other reference for those.

For the float gadget I used this as a reference: https://github.com/matter-labs/sapling-crypto/blob/master/src/circuit/float_point.rs